components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package io.gitpod.jetbrains.remote
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.application.runInEdt
+import com.intellij.openapi.diagnostic.thisLogger
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.wm.ToolWindowManager
+import com.intellij.openapi.wm.ex.ToolWindowManagerListener
+import com.intellij.remoteDev.util.onTerminationOrNow
+import com.intellij.util.application
+import com.jetbrains.rd.util.lifetime.Lifetime
+import com.jetbrains.rdserver.terminal.BackendTerminalManager
+import io.gitpod.jetbrains.remote.utils.Retrier.retry
+import io.gitpod.supervisor.api.Status.*
+import io.gitpod.supervisor.api.StatusServiceGrpc
+import io.gitpod.supervisor.api.TerminalOuterClass
+import io.gitpod.supervisor.api.TerminalServiceGrpc
+import io.grpc.stub.ClientCallStreamObserver
+import io.grpc.stub.ClientResponseObserver
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.DelicateCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.guava.asDeferred
+import kotlinx.coroutines.launch
+import org.jetbrains.plugins.terminal.TerminalToolWindowFactory
+import org.jetbrains.plugins.terminal.TerminalView
+import java.util.concurrent.CompletableFuture
+
+@Suppress("UnstableApiUsage", "EXPERIMENTAL_IS_NOT_ENABLED", "OPT_IN_IS_NOT_ENABLED")
+@OptIn(DelicateCoroutinesApi::class)
+class GitpodTerminalService(private val project: Project) : Disposable {
+    private val lifetime = Lifetime.Eternal.createNested()
+    private val terminalView = TerminalView.getInstance(project)
+    private val terminalServiceFutureStub = TerminalServiceGrpc.newFutureStub(GitpodManager.supervisorChannel)
+    private val statusServiceStub = StatusServiceGrpc.newStub(GitpodManager.supervisorChannel)
+    private val backendTerminalManager = BackendTerminalManager.getInstance(project)
+
+    override fun dispose() {
+        lifetime.terminate()
+    }
+
+    init {
+        if (!application.isHeadlessEnvironment) {
+            launch()
+        }
+    }
+
+    private fun launch() = GlobalScope.launch {
+        retry(3) { getTerminalToolWindowRegisteredEvent().await() }
+        val tasks = retry(3) { getSupervisorTasksList().await() }
```

Comment:
What is the value if retry fails 3 times? It seems code below assume only successful case? Why is it better than retrying forever and handling cancellation?

Reply:
A forever retrial is better, indeed.
I've updated the code, removing those `retry`.